gax/src/main/java/com/google/api/gax/rpc/StubSettings.java

```diff
@@ -177,11 +189,29 @@ protected Builder(StubSettings settings) {
       this.internalHeaderProvider = settings.internalHeaderProvider;
       this.clock = settings.clock;
       this.endpoint = settings.endpoint;
+      this.quotaProjectID = settings.quotaProjectID;
       this.streamWatchdogProvider = settings.streamWatchdogProvider;
       this.streamWatchdogCheckInterval = settings.streamWatchdogCheckInterval;
       this.tracerFactory = settings.tracerFactory;
     }
 
+    /** Get Quota Project ID from Client Context * */
+    private static String getQuotaProjectIDFromClientContext(ClientContext clientContext) {
+      if (clientContext.getQuotaProjectID() != null) {
+        return clientContext.getQuotaProjectID();
+      }
+      if (clientContext.getCredentials() instanceof QuotaProjectIdProvider) {
+        return ((QuotaProjectIdProvider) clientContext.getCredentials()).getQuotaProjectId();
+      }
+      if (clientContext.getHeaders().containsKey(QUOTA_PROJECT_ID_HEADER_KEY)) {
+        return clientContext.getHeaders().get(QUOTA_PROJECT_ID_HEADER_KEY);
+      }
+      if (clientContext.getInternalHeaders().containsKey(QUOTA_PROJECT_ID_HEADER_KEY)) {
```

Comment:
What is the difference between headers and internal headers? And is it possible for quota project to come in on either one?

Also, what is the use case for pulling it from the header? I think maybe it is not necessary to detect this (@broady ?)

Reply:
quota project id can be gotten from header and internal header. double confirm with @chingor13 

but I am not sure the use case.  @broady

Reply:
I believe the internal header provider is reserved for gapic and handwritten layers to inject headers. The general header provider is to allow library users to specify arbitrary extra headers.

It's not desired for library users to set the quota project via header provider, but previously that was the only mechanism for doing so. We should have an explicit strategy for how to handle that case.

Reply:
Thanks for clarifying.

Reply:
It seems ok to support this then, for backwards compatibility. And also to ensure any quota project behavior stays consistent regardless of how it's set.